routing/road_index.hpp

```diff
@@ -0,0 +1,134 @@
+#pragma once
+
+#include "routing/joint.hpp"
+
+#include "coding/reader.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "std/cstdint.hpp"
+#include "std/unordered_map.hpp"
+#include "std/utility.hpp"
+#include "std/vector.hpp"
+
+namespace routing
+{
+class RoadJointIds final
+{
+public:
+  Joint::Id GetJointId(uint32_t segId) const
+  {
+    if (segId < m_jointIds.size())
+      return m_jointIds[segId];
+
+    return Joint::kInvalidId;
+  }
+
+  void AddJoint(uint32_t segId, Joint::Id jointId)
+  {
+    if (segId >= m_jointIds.size())
+      m_jointIds.insert(m_jointIds.end(), segId + 1 - m_jointIds.size(), Joint::kInvalidId);
+
+    ASSERT_EQUAL(m_jointIds[segId], Joint::kInvalidId, ());
+    m_jointIds[segId] = jointId;
+  }
+
+  template <typename F>
+  void ForEachJoint(F && f) const
+  {
+    for (uint32_t segId = 0; segId < m_jointIds.size(); ++segId)
+    {
+      Joint::Id const jointId = m_jointIds[segId];
+      if (jointId != Joint::kInvalidId)
+        f(segId, jointId);
+    }
+  }
+
+  size_t GetSize() const { return m_jointIds.size(); }
+
+  template <class TSink>
+  void Serialize(TSink & sink) const
+  {
+    WriteToSink(sink, static_cast<Joint::Id>(m_jointIds.size()));
+    for (Joint::Id jointId : m_jointIds)
+      WriteToSink(sink, jointId);
+  }
+
+  template <class TSource>
+  void Deserialize(TSource & src)
+  {
+    Joint::Id const jointsSize = ReadPrimitiveFromSource<Joint::Id>(src);
+    m_jointIds.reserve(jointsSize);
+    for (Joint::Id i = 0; i < jointsSize; ++i)
+    {
+      Joint::Id const jointId = ReadPrimitiveFromSource<Joint::Id>(src);
+      m_jointIds.emplace_back(jointId);
+    }
+  }
+
+private:
+  // Joint ids indexed by segment id.
+  // If some segment id doesn't match any joint id, this vector contains Joint::kInvalidId.
+  vector<Joint::Id> m_jointIds;
+};
+
+class RoadIndex final
+{
+public:
+  void Import(vector<Joint> const & joints);
+
+  void AddJoint(RoadPoint rp, Joint::Id jointId)
```

Comment:
RoadPoint is pair of int, 8 bytes.
For so small class passing by value is more effective then by ref.